[요구사항 6] 하리 제출합니다 🌕

LAH1203 opened the pull request:
와 도대체 이게 얼마만인지 모르겠네요 감격,, 🙊

한 가지 양해의 말씀을 드리자면 제가 이번 주 내내 여행을 하느라 오늘에서야 이 작업을 시작해서 지금 코드가 많이 더럽습니다 😂
기간을 지키는 게 더 좋을 듯 싶어 대충이라도 어기여차 만들어서 제출합니다,, 많은 피드백 부탁드려요!

### [데모 페이지](https://lah1203.github.io/2022-looking-for-the-sound-of-an-object/)

<br />

### 참고

이번에 recoil 썼어요! 시간 없어서 대충 썼음다.. 난생 처음 봐서 아직 뭐가 뭔지 모르는 응애이니 귀엽게 봐주세요 🫣

<br />

### 컴포넌트 분리 기준

저는 이번에는 기존에 사용하던 것과는 다른 방식으로 컴포넌트들을 나누고 싶어서 새로운 방식을 적용해봤어요!

우선 페이지에서는 특정 컴포넌트에 대한 로직 없이 컴포넌트들을 엮어 페이지로 보여주는 부분만 존재합니다.

```ts
function MainPage() {
  return (
    <S.VendingMachine>
      <Menu />
      <GuideMessage />
      <Change />
      <Dispenser />
    </S.VendingMachine>
  );
}
```

그리고 위에서 볼 수 있다시피 `Menu`, `GuideMessage`, `Change`, `Dispenser`로 컴포넌트를 각자 역할에 맞게 나누어보았습니당~!

<img width="300" alt="image" src="https://user-images.githubusercontent.com/57928612/174262687-34ee1c3e-dce7-4986-b12a-b53003c81a5e.png">

그리고 가이드메시지, 잔돈, 현재 디스펜서의 상태(비어 있는지, 만들고 있는 중인지)는 recoil을 사용하여 전역 상태로 관리해보았는데요! 사실 전역으로 관리할 필요 없이 props로만 관리해도 될 것이 보이긴 하는데, recoil을 한 번 사용해보고 싶어서 우선 현재 가진 모든 상태를 전역 상태로 만들어보았습니다. 이 부분도 언제든 피드백 환영입니다! 😆

<br />

*하다 보니 이전에 자바스크립트로만 짜던 건 전부 버리고 새로 시작한 셈이 되어버렸네요ㅎㅎ; 바빠서 어쩔 수 없었습니다.. 이해해주세요...*

그럼 잘 부탁드립니다! 감사합니다 🤗

LAH1203 committed:
docs: 6단계 요구사항 작성

LAH1203 committed:
refactor: 각 컴포넌트 네이밍 변경

LAH1203 committed:
feat: 금액 투입 시 input clear 기능 추가

LAH1203 committed:
feat: 장바구니 기능 구현

LAH1203 committed:
refactor: if-else 구문 switch로 대체

LAH1203 committed:
style: react migration

LAH1203 committed:
chore: gh-pages 설치

LAH1203 committed:
refactor: 잔돈 충전 부분 분리

LAH1203 committed:
refactor: change output 분리

LAH1203 committed:
refactor: dispenser, guide message 분리

LAH1203 committed:
style: guiding → guideMessage로 이름 변경

moonheekim0118 commented on src/components/MenuButton/index.type.ts:
```diff
@@ -0,0 +1,7 @@
+export interface Props {
+  drink: {
+    name: string;
+    price: number;
+    ingredients: String[];
+  };
+}
```
해당 타입은 drink 객체의 구조/타입이 바뀌면 언제든지 바뀌어야 할 것 같아요!
그런데, drink객체의 구조/타입이 바뀌었을 때, MenuButton 컴포넌트 에서 drink 객체를 props로 받고있는지 개발자가 바로 알 수 있을까요🤔? 만약 다른 컴포넌트에서도 drink 객체를 props로 받아서 사용하고 있다면..? 그래서 타입 정의를 여러군데에서 했다면..? 한번에 바꾸기 조금 어려울 것 같아요!
저의 개인적인 의견으로는! drink 객체의 type 은 언제 어디서든지 재활용 될 수 있고, 
변경이 요구될 때 한번에 변경되는 것이 휴먼에러를 방지할 것 같기 때문에 types 라는 공통 폴더를 만들어서 공통적으로 한번에 관리해주는게 어떤지 제안드립니다!

LAH1203 replied:
오 너무 좋아요! 현재는 drink 객체를 여기에서만 정의하여 사용하고 있어서 이렇게 만들었었네요 😱
바로 피드백 반영해서 Drink type 정의해보도록 하겠습니다..❤️ [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/f1521fdac72ddca7bba4049f64e53d2fd66479f4)

moonheekim0118 commented on src/constants/dispenser.ts:
```diff
@@ -0,0 +1,6 @@
+const DISPENSER_STATE = {
+  EMPTY: 'empty',
+  MAKING: 'making',
+};
```
```suggestion
const DISPENSER_STATE = {
  EMPTY: 'empty',
  MAKING: 'making',
} as const; 
```

요고 하리 아래에는 적용했는데 여기는 안한거보면,  깜빡한 것 같은데! 아니면, 이 상수 객체에만 as const 를 안붙인 이유가 따로 있을까요?

LAH1203 replied:
제가 어디에 했나 생각이 안 나서 보고 왔습니다 🤣 theme 정의할 때 썼었네용,,
이 부분은 제가 특별한 기준 없이 어디에는 쓰고 어디에는 안 쓰고 했던 것 같아요!
알려주셔서 감사합니다 :) 앞으로 상수 정의할 때는 붙이는 방향으로 기준을 세워볼게용 😆 [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/8790e28003247879107d1269c0acfe26480ef0ea)

moonheekim0118 commented on src/constants/menu.ts:
```diff
@@ -0,0 +1,24 @@
+const menu = [
+  {
+    name: '에스프레소',
+    price: 1000,
+    ingredients: ['에스프레소'],
+  },
+  {
+    name: '아메리카노',
+    price: 1500,
+    ingredients: ['에스프레소', '뜨거운 물'],
+  },
+  {
+    name: '카페라떼',
+    price: 2000,
+    ingredients: ['우유', '에스프레소'],
+  },
+  {
+    name: '우유',
+    price: 1000,
+    ingredients: ['우유'],
+  },
+];
```
저 역시 음료 정보를 상수로 관리해주어서 컴포넌트에서 바로 사용했는데요, 하리가 recoil을 멋지게 도입해서 문득 든 생각이에요.
만약, 해당 어플리케이션을 서버와 연결하여 관리한다고 하면, 음료 정보 역시 서버에서 가져오게 될 정보가 아닐까 싶어요!
그렇게 된다면 음료 정보는 상수에서 바로 가져오는 형태가 아니라, 전역 상태에 저장해서 관리하는게 더 확장성이 좋지 않을까 싶어요. (나중에 서버와 연동할 때, 바로 갈아끼우면 되니까?)

그런데 이는 어플리케이션 성격에 따라서 (클라이언트에서 음료 정보 관리를 하느냐, 서버에서 음료 정보 관리를 하느냐) 달라질 것 같아요 😭 
하리는 어떻게 생각하시나요?

LAH1203 replied:
흠 사실 저도 이 부분에 대해서 서버에 요청하는 방식을 고려해서 전역에 저장할까도 생각해봤는데요! 현재는 서버나 msw가 없는 상태여서 아직은 없어도 되지 않을까 생각이 들어서 우선 상수로 사용했습니당
그래서 현재는 클라이언트에서 음료 정보를 관리하는 방향으로 만들어졌는데, 만약에 추후 임시 서버나 msw를 도입하게 된다면! 그 때 서버에서 받아와 전역에 저장하는 형태로 만들어도 좋지 않을까 싶어요. (사실 아직 recoil을 잘 몰라서 어떻게 서버에 요청해서 저장하는지를 좀 더 알아봐야 할 것 같기도 하고요 🤣)
좋은 의견 감사합니다 😆

moonheekim0118 commented on src/recoil/states.ts:
```diff
@@ -0,0 +1,20 @@
+import { atom } from 'recoil';
+
+import { DISPENSER_STATE } from 'constants/dispenser';
+
+const guideState = atom({
+  key: 'guideState',
+  default: '음료를 골라주세요.',
+});
+
+const changeState = atom({
+  key: 'changeState',
+  default: 0,
+});
+
+const dispenserState = atom({
+  key: 'dispenserState',
+  default: DISPENSER_STATE.EMPTY,
+});
+
+export { guideState, changeState, dispenserState };
```
![](http://file3.instiz.net/data/file3/2018/01/29/b/2/a/b2aa8e9727ab9d539ec4ae63fb5ced22.gif)
recoil 이렇게 쓰는거구나..!
새로운 기술을 샤샤샥 잘 도입하다니 너무 멋져요 ><

moonheekim0118 replied:
궁금해서 찾아보니 recoil 로 관리하는 전역 상태에도 아래와 같이 타입을 정의할 수 있네요!
타입스크립트 + 리코일로 더 안전한 전역 상태 관리를 해보는건 어떨까요! 
```tsx

const changeState = atom<number>({
  key: 'changeState',
  default: 0,
});


```

LAH1203 replied:
오호 그대는 천재..? 너무 좋은데요!? 감사합니다 😘
[commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/65719795226392e1c3f5cbd30de7a0888d44fd96)

moonheekim0118 commented on src/styles/GlobalStyle.ts:
```diff
@@ -0,0 +1,20 @@
+import { createGlobalStyle } from 'styled-components';
+import reset from 'styled-reset';
+
+const GlobalStyle = createGlobalStyle`
+  ${reset}
```
![](https://coinpan.com/files/attach/images/198/084/791/183/f45397df5adddde6cd73e6265711d81a.gif)
reset 과 styled-reset에 대해서 설명해주실 수 있으실까용 ?

LAH1203 replied:
보통 우리가 태그를 사용하다보면 기본적으로 주어진 스타일링(ex: margin 👿)이 우리를 거슬리게 했던 경험이 몇 번 있지 않나요? css를 사용할 때는 `reset.css`란 파일에서 그러한 스타일링들을 초기화시켜주고는 했었는데, `styled-reset`에서는 이를 'reset'이라는 이름으로 가져와서 사용할 수 있도록 지원합니다. 쓰기 편해서 사용하고 있어용 헤헤 😄

moonheekim0118 commented on src/components/MenuButton/index.tsx:
```diff
@@ -0,0 +1,53 @@
+  const [dispenser, setDispenser] = useRecoilState(dispenserState);
+
+  const orderMenu = () => {
+    if (dispenser === DISPENSER_STATE.MAKING) {
+      alert('음료 제조가 끝난 후 주문해주세요.');
+      return;
+    }
+
+    if (price > change) {
+      alert('돈이 부족합니다.');
+      return;
+    }
+
```
try-catch 구문을 써서 깔끔하게 한번에 예외 처리를 할 수 있지 않을까요? 🤔

LAH1203 replied:
오 어차피 똑같이 alert 보여주고 return 하니 그렇게 하면 더 깔끔할 것 같네요 😊
피드백 반영해서 수정해보겠습니당 [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/a767021e4ba8a2e376d222823be29a4104293953)

moonheekim0118 commented on src/components/MenuButton/index.tsx:
```diff
@@ -0,0 +1,53 @@
+    setGuideMessage('음료 제조를 시작합니다.');
+    setDispenser(DISPENSER_STATE.MAKING);
+
+    ingredients.forEach((ingredient, index) => {
+      setTimeout(() => {
+        setGuideMessage(`${ingredient}가(이) 나옵니다.`);
+      }, 1000 * (index + 1));
+    });
+
+    setTimeout(() => {
+      setGuideMessage(`${name} 제조가 완료되었습니다.`);
+      setDispenser(DISPENSER_STATE.EMPTY);
+    }, 1000 * (ingredients.length + 1));
+  };
```
이 setTimeout 들은 모두 실행된 후, clear가 되고 있지 않아서 성능상 좋지 못할 것 같아요. 🥺
한번 개선해볼까요?

LAH1203 replied:
제가 알고 있던 바로는 clearTimeout을 하는 경우는 setTimeout을 취소하는 경우에 해당하는데 혹시 잘못 알고 있을 수도 있어서 관련해서 여러 글을 찾아봤어용

[글 1 - stackoverflow](https://stackoverflow.com/questions/7391567/when-using-settimeout-do-you-have-to-cleartimeout)
[글 2 - stackoverflow](https://stackoverflow.com/questions/32580940/is-cleartimeout-necessary-after-settimeout-with-0ms)
[setTimeout - MDN](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout)
[clearTimeout - MDN](https://developer.mozilla.org/en-US/docs/Web/API/clearTimeout)

- 위의 글들에서는 clearTimeout을 setTimeout 취소할 때 호출한다는 얘기를 볼 수 있었습니다.

[글 3 - stackoverflow](https://stackoverflow.com/questions/31503110/will-a-settimeout-keep-a-function-from-being-garbage-collected)

- 위의 글에서는 setTimeout이 완료되면 자동으로 가비지 컬렉터에 의해 수집된다는 얘기를 볼 수 있었습니다.

setInterval 같은 경우에는 clearInterval이 필요한데, setTimeout에는 필요하지 않다는 글이 많았습니당,, 그래서 저는 clear하지 않아도 된다고 생각했는데, 혹시 호프는 다른 의견이시라면 호프의 의견도 들어보고 싶네요! 😉

moonheekim0118 commented on src/components/MenuButton/index.tsx:
```diff
@@ -0,0 +1,53 @@
+      <S.Label>
+        {name}
+        <br />({price}원)
+      </S.Label>
+      <S.Button type="button" onClick={orderMenu} canBuy={price <= change} />
```
Button styled-component 입장에서 `canBuy`라는 props 명을 이해할 수 있을까요?
스타일링 컴포넌트가 도메인 (buy) 정보를 너무 깊게 알고 있는 것 같아요!
만약 해당 styled-component 가 재사용되거나, 도메인의 성격이 바뀐다면 canBuy 는 올바르지 못할 것 같은데,
조금 더 범용성 있는 네이밍을 써볼까요?

LAH1203 replied:
아래 피드백과 같은 맥락이라고 생각해서 아래에 남겨볼게요!

moonheekim0118 commented on src/components/MenuButton/index.styled.ts:
```diff
@@ -0,0 +1,29 @@
+  font-size: 15px;
+`;
+
+const Button = styled.button`
+  width: 100%;
+  height: 20px;
+  border-radius: 50px;
+
+  ${({ canBuy }: { canBuy: boolean }) => css`
+    background: ${({ theme: { colors } }) =>
+      canBuy ? colors.brown001 : colors.gray002};
+  `}
```
button html element 를 사용하고 있으니, 우리 html element 가 주는 네이티브 속성을 최대한 활용해보아요!

아예 disabled attribute 를 넘겨주고  , disabled 속성 스타일링을 정의해주는건 어떨까요..?!
```suggestion
const Button = styled.button`
  width: 100%;
  height: 20px;
  border-radius: 50px;
   background: ${({ theme: { colors } }) =>
      colors. brown001};

 &: disabled { 
        background: ${({ theme: { colors } }) =>
      colors.gray002};
 }
```

사용되는 곳

```tsx

<Button disabled={ //... true or false }/>
```

LAH1203 replied:
오 너무 좋은데요?!?! 왜 하면서는 그 생각을 못했을까여,, 🥲
바로 수정하겠습니당 친절한 예시 코드까지,, 캄사합니둥 🥰 [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/8c2a1be1173225d62443316721c0b6615b4b5e1f)

moonheekim0118 commented on src/components/Dispenser/index.styled.ts:
```diff
@@ -0,0 +1,32 @@
+  height: 10px;
+  border: none;
+  border-radius: 50%;
+
+  ${({ making }: { making: boolean }) => css`
+    background: ${({ theme: { colors } }) =>
+      making ? colors.red001 : colors.green001};
+  `}
```
저는 대체로 boolean 값을 가지고 있는 변수는 is-* 프리픽스를 사용해서 t/f 값인지 명확히 알려주려구 하는데, 어떠신가요 ? ㅎㅎ

LAH1203 replied:
좋은 생각이에요! 저도 보통 boolean 값에는 is prefix를 쓰는데 여긴 왜 이렇게 썼는지 살짝 의문이네요 ㅎㅎㅠ
피드백 반영 [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/720dd79c95a80c8ce8edd27274261980b4cc0dbc)

moonheekim0118 commented on src/components/ChangeOutput/index.tsx:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  return (
+    <S.ChangeContainer>
+      <S.Change>{change.toLocaleString('ko-KR')}</S.Change>
+      <S.ReturnButton type="button" onClick={returnChange}>
```
`toLocaleString('ko-KR')` 은 유틸함수로 분리하는게 어떨까요?
UI 요구사항이 바뀌면 변경되어야 할 부분이라서, 컴포넌트 외부에서 따로 관리해주는게 좋을 것 같아요!

LAH1203 replied:
호 전에 이와 비슷한 걸 피드백으로 받은 적이 있었던 것 같은데 항상 빼먹게 되네요 💦
저도 같은 생각이에요! 유틸로 분리해볼게요 :) [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/6ed6dba41fc119961fd484e42c1f16463030453d)

moonheekim0118 commented on src/components/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,37 @@
+
+  const chargeChange = (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!(e.target instanceof HTMLFormElement)) return;
+
```
![](https://mblogthumb-phinf.pstatic.net/MjAxODAzMjZfMTM2/MDAxNTIyMDQ4MjQ5NDIx.2axZfCxRJIDQf6hzeUwG0H62qbWlPAmtslEeGqrBTl4g.H_ks_4jYefCplaJFzZrYso19nYV5v--0c76d8mbUH3Eg.GIF.ok00urb/%EF%BF%BD%EF%BF%BD%EF%BF%BD-%EF%BF%BD%EF%BF%BD%EF%BF%BD-121.gif?type=w800)
이게 바로 타입가드..?!

moonheekim0118 commented on src/components/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,37 @@
+    if (!(e.target instanceof HTMLFormElement)) return;
+
+    const money = (e.target.elements[0] as HTMLInputElement).valueAsNumber;
+
+    if (!money || money % 500 !== 0) {
+      alert('500원 단위로 입력해주세요.');
+      return;
+    }
```
이런 validation은 따로 util 로 분리하면 관리하기 쉽지 않을까요~?

LAH1203 replied:
사실 validate로 빼기엔 너무 간단해서 우선 그냥 뒀는데요!
그래도 피드백 받았으니 분리하는 게 좋겠네요!
분리 완료 🌟 [commit](https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18/commits/a409f080097e79d8ffa2cee787db7dd236897b12)

moonheekim0118 commented on src/react-app-env.d.ts:
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
이건 혹시 왜 필요한가요...? 🧐

LAH1203 replied:
흠 그냥 cra typescript template으로 만들었었는데 이런게 생겼나보네요..? 궁금해서 찾아봤습니다 🥸

[참고 - 공식](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html)

> Triple-slash references instruct the compiler to include additional files in the compilation process.

`react-scripts`에 있는 내용(타입 선언)을 가져와서 사용하도록 하는 것인가봅니다. (자바스크립트의 import 같은 녀석인가..)

moonheekim0118 commented on public/manifest.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "short_name": "R.B.V.M",
+  "name": "React Beverage Vending Machine",
+  "icons": [
+    {
+      "src": "favicon.ico",
+      "sizes": "64x64 32x32 24x24 16x16",
+      "type": "image/x-icon"
+    },
+    {
+      "src": "logo192.png",
+      "type": "image/png",
+      "sizes": "192x192"
+    },
+    {
+      "src": "logo310.png",
+      "type": "image/png",
+      "sizes": "310x310"
+    }
+  ],
+  "start_url": ".",
+  "display": "standalone",
+  "theme_color": "#000000",
+  "background_color": "#ffffff"
+}
```
로고까지 바꿔서 매니페스트 파일에도 적용해주었군요! 넘 꼼꼼하고 멋져요! 👍 👍

moonheekim0118 commented on src/components/ChangeInput/index.tsx:
```diff
@@ -0,0 +1,37 @@
+    setGuideMessage('투입되었습니다.');
+  };
+
+  return (
+    <S.ChangeInputForm onSubmit={chargeChange}>
+      <input type="number" />
+      <button>투입</button>
+    </S.ChangeInputForm>
+  );
+}
+
+export default ChangeInput;
```
컴포넌트 네이밍이 `ChangeInput` 인데 사실상 컴포넌트는 Form 역할을 하고 있는게 조금 헷갈리는 것 같아요! 😮
물론 Output- Input으로 네이밍을 통일해주어서 깔끔하지만! 
저는 최대한 해당 컴포넌트에서 실질적으로 사용되는 Element 네이밍과 컴포넌트 네이밍을 맞추는 편이거든요! 
하리는 어떻게 생각하시나요?!

LAH1203 replied:
흠 사실 저도 컴포넌트 최상단 엘리먼트 네이밍과 컴포넌트 네이밍을 맞추는 편이긴 합니당,,
사실 이렇게 한 것은 호프가 말해주셨다시피 Input-Output으로 네이밍을 통일해주기 위함도 있었지만, 해당 컴포넌트에 form이 존재한다는 건 그 컴포넌트에 대해 설명하는 데 큰 비중을 가지고 있지 않기 때문이에요. 저는 원래 event handler로 전달해주는 함수 이름을 `on${EventName}${Element}`로 만들어주고는 했는데, 이렇게 하면 해당 함수가 무슨 동작을 하는지 알기 힘들다고 하더라고요. 저도 그 의견에 공감해서 이젠 함수 이름에 그 역할을 담으려고 하고 있는데, 컴포넌트도 마찬가지라고 생각했어요. '네가 어떻게 생겼는지는 궁금하지 않아. 역할만 알려줘!' 느낌으로? 그래서 여기에서도 그렇게 사용해봤습니당 😌

moonheekim0118 reviewed (changes requested):
![](https://mblogthumb-phinf.pstatic.net/MjAxOTEwMDhfMTM4/MDAxNTcwNDg4NzY0NDMw.UH-nzGRnaT5N-ZjNYPpQDbBvgh82Grnhhzd5NElQfu8g.ssoSG3u5dmz7ImfIvhn5MH8i1QaFG3N8un70q7B6EW8g.GIF.kwoun486/16C96C98-09FF-4710-A89B-70D057379CFB.gif?type=w800)

안녕하세요  ❤️ 하 ❤️ 리  ❤️    방학 즐겁게 보내구 계신지요~?
하리의 자판기를 리뷰하게 되어서 영광입니다!
특히나 recoil을 과감하게 시도해본 도전이 너무나 멋져요 👍 👍  하리 덕분에 recoil에 대해 배울 수 있었기에 고맙다고 말해주고 싶습니다 흐흐 
역시나 타입스크립트를 너무나 잘 사용해주고 있고! 컴포넌트 분리도 너무 깔끔한 것 같아요.
전역 상태 라이브러리를 사용하니까 props drilling 이 제거되어서 컴포넌트가 굉장히 독립적이어지네요.
아주 사소한 몇가지 코멘트랑 질문 의견 남겨놨으니, 편안하게 답해주시면 감사하겠읍니다 😃 ❤️ 
수고했어요! 👍 👍 👍

LAH1203 committed:
refactor: Drink 객체 타입 정의

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901065941

LAH1203 committed:
refactor: 상수에 as const 적용

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901066171

LAH1203 committed:
fix: 상수에 as const 도입에 따른 타입 변경

LAH1203 committed:
feat: recoil state에 타입 정의

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901072523

LAH1203 committed:
refactor: 예외 처리 try-catch로 변경

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901067423

LAH1203 committed:
refactor: canBuy → disabled로 props 이름 변경

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901067850
- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901068289

LAH1203 committed:
refactor: making → isMaking으로 props 이름 변경

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901068524
- boolean 값이므로 is-* prefix 사용

LAH1203 committed:
refactor: number format 로직 util로 분리

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901068798

LAH1203 committed:
style: utils format 파일명 변경

LAH1203 committed:
refactor: change input validate util로 분리

- https://github.com/woowacourse-study/2022-looking-for-the-sound-of-an-object/pull/18#discussion_r901069300
